src/dos/dos_code_page.cpp

```diff
@@ -0,0 +1,2223 @@
+
+struct ParserResult {
+	// A general return code; 'false' if the file can't be parsed at all
+	bool status_ok = true;
+
+	// If the reson of parsing failure was an unsupported UPX compression
+	bool unsupported_cpx_file = false;
+
+	bool found_printer_fonts = false;
+	bool found_screen_fonts  = false;
+
+	// Fonts found but not necessarily extracted
+	std::set<uint16_t> found_ega_fonts   = {};
+	std::set<uint16_t> found_lcd_fonts   = {};
+	std::set<uint16_t> found_other_fonts = {};
+
+	// These are fonts which were extracted
+	ScreenFonts extracted_ega_fonts   = {};
+	ScreenFonts extracted_lcd_fonts   = {};
+	ScreenFonts extracted_other_fonts = {};
+
+	ParserResult(const bool status_ok) : status_ok(status_ok) {}
+	ParserResult() = default;
+};
+
+class CpiParser : public CpiReader {
+public:
+	CpiParser(const std::vector<uint8_t>& content, const std::string& name_for_log)
+	        : CpiReader(content),
+	          name_for_log(name_for_log)
+	{}
+
+	ParserResult GetFonts(const uint16_t code_page_filter = 0);
+
+private:
+	const std::string name_for_log = {};
+
+	FileFormat file_format = FileFormat::Unknown;
+
+	// Offsets needed by the screen font format 2
+	struct FontOffsets {
+		std::optional<uint32_t> start_8x16 = {};
+		std::optional<uint32_t> start_8x14 = {};
+		std::optional<uint32_t> start_8x8  = {};
+	};
+
+	FontOffsets font_offsets = {};
+
+	struct Found {
+		std::set<std::string> device_names                    = {};
+		std::set<std::pair<std::string, uint16_t>> code_pages = {};
+	};
+
+	Found found = {};
+
+	// Initialization, file format detection
+
+	void InitParser();
+	void DetectFileFormat(const uint8_t id_byte, const std::string& id_string);
+
+	// Common/generic helper functions
+
+	bool IsFormatMsDos() const
+	{
+		return file_format == FileFormat::MsDos;
+	}
+	bool IsFormatWinNt() const
+	{
+		return file_format == FileFormat::WinNt;
+	}
+	bool IsFormatDrDos() const
+	{
+		return file_format == FileFormat::DrDos;
+	}
+
+	static uint32_t ConvertOffset(const uint32_t value);
+
+	static bool IsCodePageMatching(const CodePageEntryHeader& header,
+	                               const uint16_t code_page_filter);
+
+	static bool IsDeviceNameValid(const CodePageEntryHeader& header);
+
+	static bool IsEgaCodePage(const CodePageEntryHeader& header);
+	static bool IsLcdCodePage(const CodePageEntryHeader& header);
+
+	static bool IsPrinterCodePage(const CodePageEntryHeader& header);
+	static bool IsScreenCodePage(const CodePageEntryHeader& header);
+
+	void StoreFontType(const CodePageEntryHeader& header, ParserResult& result);
+
+	// Concrete screen font extraction
+
+	std::optional<ScreenFont> GetScreenFontFormat1(const uint16_t num_fonts,
+	                                               const uint16_t total_size);
+	std::optional<ScreenFont> GetScreenFontFormat2(const uint16_t num_fonts,
+	                                               const uint16_t total_size);
```

Comment:
Can't just use `int`s?

Reply:
These values are taken straight from the binary data format, I would prefer to keep the original integer width.